File: Problems 0500-0599/543. Diameter of Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        return self.diameter(root)[1]

    def diameter(self, node):
        if not node:
            return [-1, 0]

        onePathL, twoPathL = self.diameter(node.left)
        onePathR, twoPathR  = self.diameter(node.right)

        return [1 + max(onePathL, onePathR), max(2 + onePathL + onePathR, max(twoPathL, twoPathR))]

File: Problems 0500-0599/543. Diameter of Binary Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 class Solution {
    public:
        int diameterOfBinaryTree(TreeNode* root) {
            int res = 0;
            getDiameter(root, res);
            return res;
        }
    
        int getDiameter(TreeNode* node, int& res) {
            if (!node) return 0;
    
            int leftDepth = getDiameter(node->left, res);
            int rightDepth = getDiameter(node->right, res);
            
            res = max(res, leftDepth + rightDepth);
            return 1 + max(leftDepth, rightDepth);
        }
    };